client/service/src/builder.rs

```diff
@@ -900,14 +901,55 @@ pub fn build_network<TBl, TExPool, TImpQu, TCl>(
 		config.announce_block,
 	);
 
+	// TODO: Normally, one is supposed to pass a list of notifications protocols supported by the
+	// node through the `NetworkConfiguration` struct. But because this function doesn't know in
+	// advance which components, such as GrandPa or Polkadot, will be plugged on top of the
+	// service, it is unfortunately not possible to do so without some deep refactoring. To bypass
+	// this problem, the `NetworkService` provides a `register_notifications_protocol` method that
+	// can be called even after the network has been initialized. However, we want to avoid the
+	// situation where `register_notifications_protocol` is called *after* the network actually
+	// connects to other peers. For this reason, we delay the process of the network future until
+	// the user calls `NetworkStarter::start_network`.
+	//
+	// This entire hack should eventually be removed in favour of passing the list of protocols
+	// through the configuration.
+	//
+	// See also https://github.com/paritytech/substrate/issues/6827
+	let (network_start_tx, network_start_rx) = oneshot::channel();
+
 	// The network worker is responsible for gathering all network messages and processing
 	// them. This is quite a heavy task, and at the time of the writing of this comment it
 	// frequently happens that this future takes several seconds or in some situations
 	// even more than a minute until it has processed its entire queue. This is clearly an
 	// issue, and ideally we would like to fix the network future to take as little time as
 	// possible, but we also take the extra harm-prevention measure to execute the networking
 	// future using `spawn_blocking`.
-	spawn_handle.spawn_blocking("network-worker", future);
+	spawn_handle.spawn_blocking("network-worker", async move {
+		if network_start_rx.await.is_err() {
+			debug_assert!(false);
+			log::warn!(
+				"The NetworkStart returned as part of `build_network` has been silently dropped"
+			);
+			// This `return` might seem unnecessary, but we don't want to make it look like
+			// everything is working as normal even though the user is clearly misusing the API.
+			return;
+		}
+
+		future.await
+	});
+
+	Ok((network, network_status_sinks, system_rpc_tx, NetworkStarter(network_start_tx)))
+}
 
-	Ok((network, network_status_sinks, system_rpc_tx))
+/// Object used to start the network.
+#[must_use]
+pub struct NetworkStarter(oneshot::Sender<()>);
+
+impl NetworkStarter {
+	/// Start the network. Call this after all sub-components have been initialized.
+	///
+	/// > **Note**: If you don't call this function, the networking will stay inhert.
```

Comment:
```suggestion
	/// > **Note**: If you don't call this function, the networking will stay inert.
```

Is this what you mean? It might be better to say something simpler like 

```suggestion
	/// > **Note**: If you don't call this function, the networking will not work.
```
so that it's easier to understand.